designs/2025-define-plugin/README.md

```diff
@@ -0,0 +1,628 @@
+    recommended: {
+      rules: Object.values(rules).filter((rule) => rule.meta.docs.recommended),
+    },
+    recommendedTypeChecked: {
+      rules: Object.values(rules).filter(
+        (rule) =>
+          rule.meta.docs.recommended && rule.meta.docs.requiresTypeChecking
+      ),
+    },
+    strict: {
+      rules: Object.values(rules).filter(
+        (rule) => rule.meta.docs.recommended === "strict"
+      ),
+    },
+    strictTypeChecked: {
+      rules: Object.values(rules).filter(
+        (rule) =>
+          rule.meta.docs.recommended === "strict" &&
+          rule.meta.docs.requiresTypeChecking
+      ),
+    },
+  },
+  meta,
+  rules,
+});
+```
+
+### Lazy-Loading Rules
+
+[ESLint core implements lazy-loaded configs with a `LazyLoadingRuleMap` and `Proxy`](https://github.com/eslint/eslint/blob/129882d2fdb4e7f597ed78eeadd86377f3d6b078/lib/config/default-config.js#L46)
+[vuejs/eslint-plugin-vue#2732](https://github.com/vuejs/eslint-plugin-vue/issues/2732) and [typescript-eslint/typescript-eslint#11029 Enhancement: Support Lazy Loading Rules](https://github.com/typescript-eslint/typescript-eslint/issues/11029) ask for plugins to lazy-load their rules too.
+Doing so would allow plugins with many dozens or more rules to avoid the cost of loading those rules unnecessarily.
+
+`definePlugin` can enable lazy-loaded rules by allowing elements in `rules` arrays to each be provided as either:
+
+- A rule object itself
+- A function that synchronously returns a rule
+
+A plugin definition that uses lazy-loaded rules could look like:
+
+```js
+import { definePlugin } from "@eslint/plugin-kit";
+import { createRequire } from "node:module";
+
+import packageData from "../package.json" with { type: "json" };
+
+const require = createRequire(import.meta.url);
+
+export const plugin = definePlugin({
+  configs: {
+    recommended: {
+      rules: {
+        "example/a": "error",
+        "example/b": "error",
+        "example/c": "error",
+        // ...
+      },
+    },
+  },
+  meta: { name: packageData.name, version: packageData.version },
+  rules: {
+    "example/a": () => require("./rules/a"),
+    "example/b": () => require("./rules/b"),
+    "example/c": () => require("./rules/c"),
+    // ...
+  },
+});
+```
+
+`definePlugin` will implement similar strategies internally to ESLint core, using [`module.createRequire()`](https://nodejs.org/api/module.html#modulecreaterequirefilename) to `require()` in both CommonJS and ECMAScript Modules packages.
+
+Note that lazy-loading rules conflicts with auto-generating `configs`.
+Knowing which rules are to be included in a config requires loading the rule itself.
+`definePlugin` will need to throw an error if any rule function is provided and `configs` is not specified.
+
+Two new `eslint-plugin-eslint-plugin` lint rules will be added:
+
+1. Enforcing using `definePlugin` to define a plugin
+2. Enforces lazy-loading rules and automates keeping `configs` up-to-date with lazy-loaded rules.
```

Comment:
I understand wanting to provide lazy-loading as an option. I'm not sure about enforcing its usage. If that's what we're going for, then why allow non-function values at all?

Reply:
As I see it, we can roughly divide plugin developers into two general zones:

* Folks who just want it to work, such as folks running through a "getting started" guide, or building small internal plugins. They probably benefit from having as approachable, clean syntax as possible over lazy-loading rules.
* Folks who want it to be as fast as possible, such as depended-upon open source repos or rigorous company-wide plugins. They probably care more about lazy-loading rules over having as clean syntax as possible.

I wanted to get the design to work ergonomically for the former, but still be able to satisfy the latter.

Reply:
Yeah. That does feel weird. I guess I'm a little stuck here: I don't see a way to do great at both the first-run experience _and_ getting lazy-loaded rules. So, asking for reviewers of this RFC: what is more important to ESLint? (or some similar question to that)

Reply:
Wouldn't removing just the enforcement of lazy-loaded rules be sufficient?  People who need or want to can still adopt lazy-loaded rules without the enforcement piece.

Reply:
I don't see lazy-loading as important for this design. People still can implement them on their own if they really want.

Reply:
I think trying to make `definePlugin()` do everything any plugin author might want out of the gate is a bit of a trap. I don't see lazy-loading as the most important part of this proposal, and I do see it as one that complicates the current design. To that end, I think leaving it off this proposal will allow us to make better progress. It's a lot easier to add new features later if we find it's important than it is to remove a feature that isn't quite working the way we want it to.